Implement post editing functionality with edit tracking

nyunja opened the pull request:
What was implemented:

## Backend Changes:

1. **Updated Post Model** (`backend/internal/models/post.go`):
   - Added `UpdatedAt *time.Time` field
   - Added `IsEdited bool` field

2. **Created Database Migration** (`backend/pkg/db/migrations/sqlite/000013_add_updated_at_to_posts.up.sql`):
   - Added `updated_at` column to Posts table

3. **Updated PostServiceInterface** (`backend/internal/service/interfaces.go`):
   - Added `UpdatePost(postID, userID int64, content string, imageData []byte, imageMimeType string) (*models.Post, error)` method

4. **Implemented UpdatePost in PostService** (`backend/internal/service/post_service.go`):
   - Added authorization check (only post owner can edit)
   - Content validation
   - Image handling (optional update)
   - Calls PostStore.UpdatePost

5. **Updated PostStoreInterface** (`backend/internal/store/interfaces.go`):
   - Added `UpdatePost(postID int64, content, imagePath string) (*models.Post, error)` method

6. **Implemented UpdatePost in PostStore** (`backend/internal/store/post_store.go`):
   - Updates post content, image, and sets `updated_at` timestamp
   - Returns updated post with author information
   - Sets `IsEdited` flag when `updated_at` is not null
   - Updated `GetPosts` and `GetPostByID` to include `updated_at` field and `is_edited` logic

7. **Added UpdatePost Handler** (`backend/internal/api/handlers/post_handler.go`):
   - Handles PUT requests to `/posts/{postId}`
   - Parses multipart form data
   - Validates user authorization and content
   - Handles optional image uploads

8. **Added Route** (`backend/internal/api/router.go`):
   - Added `PUT /posts/{postId}` route with authentication middleware

9. **Updated Mock Implementations**:
   - Added UpdatePost method to test mocks

## Frontend Changes:

1. **Added updatePost Function** (`frontend/lib/auth.js`):
   - Handles PUT requests to update posts
   - Supports content and optional image updates

2. **Enhanced PostList Component** (`frontend/components/posts/PostList.jsx`):
   - Added edit modal state management
   - Added `handleEditPost` function to open edit modal
   - Added `handleEditSubmit` function to submit edits
   - Connected Edit button in dropdown to `handleEditPost`
   - Added comprehensive edit modal with:
     - Textarea for content editing
     - File input for image updates
     - Cancel and Save buttons
     - Loading states and validation
   - Added "edited" indicator that shows when `post.is_edited` is true

## Key Features:

✅ **Edit Post Functionality**: Users can edit their own posts through a modal interface
✅ **Authorization**: Only post owners can edit their posts
✅ **Content Validation**: Empty content is not allowed
✅ **Image Updates**: Users can optionally update the post image
✅ **Edited Indicator**: Posts show "• edited" marker when they've been modified
✅ **Real-time Updates**: Edited posts immediately update in the feed
✅ **Database Migration**: Proper schema migration to add `updated_at` column
✅ **Comprehensive Error Handling**: Both backend and frontend handle errors gracefully

The edit functionality is now fully implemented and ready for testing! Users can click the three-dot menu on their posts, select "Edit", modify the content and/or image, and save the changes. The post will be marked as "edited" and the changes will be immediately visible in the feed.

nyunja committed:
feat: add post actions dropdown with delete confirmation

- Add dropdown menu to post cards with Edit, Delete, and Follow options
- Implement delete post functionality with confirmation modal
- Add deletePost API function to auth library
- Create confirmation dialog with "Delete" and "Cancel" options
- Add proper click outside handling for dropdown menus
- Style dropdown and modal consistently with app theme

New CSS variables added:
- --danger-color: #ff4444 (for delete button)
- --hover-background: #3a3a9a (for interactive elements)
- --border-color: #4a4ab0 (for borders)

Components updated:
- PostList.jsx: Added dropdown menu, delete handlers, and confirmation modal
- auth.js: Added deletePost API function
- globals.css: Added new color variables

Features:
- Click three dots on any post to open actions menu
- Delete option shows confirmation popup before deletion
- Posts are removed from feed immediately after successful deletion
- Dropdown closes when clicking outside or selecting an option
- Reduced modal overlay opacity for better UX

nyunja committed:
feat: implement post editing functionality with edit tracking

- Add updated_at column to Posts table via migration
- Add UpdatedAt and IsEdited fields to Post model
- Implement UpdatePost handler with authorization checks
- Add PUT /posts/{postId} route with auth middleware
- Update PostService and PostStore with edit capabilities
- Include edit status in post queries (GetPosts, GetPostByID)
- Add comprehensive error handling for unauthorized edits
- Update service and store interfaces with UpdatePost methods
- Add mock implementations for testing compatibility

nyunja committed:
feat: add post editing UI with edit indicators

- Add updatePost API function for PUT requests
- Implement edit modal with content and image update support
- Add Edit button to post dropdown menu
- Display "edited" indicator on modified posts
- Add real-time post updates after successful edits
- Include form validation and loading states
- Add proper error handling for edit operations
- Maintain consistent styling with existing modals

nyunja committed:
Merge branch 'feature/comments/130-develop-comment-creation-frontend' into feature/posts/222-implement-post-editing-functionality-with-edit-tracking

Murzuqisah commented on frontend/lib/auth.js:
```diff
@@ -99,6 +99,55 @@ export const fetchComments = async (postId) => {
+      formData.append('image', image);
+    }
+
+    const response = await fetch(`http://localhost:9000/posts/${postId}`, {
```
Use environment variables for the connection link and port

Siaka385 replied:
The entire frontend currently does not use environment variables for the ports in fetch requests. Updating this would require a consistent change across the whole frontend files, so  I do think it should be handled in a separate issue. This could be addressed during the final refactoring of the code but for now lets keep it like that.

nyunja replied:
Yes. I saw that too.

Murzuqisah commented on frontend/src/app/globals.css:
```diff
@@ -19,6 +19,11 @@
   /* Status Colors */
   --sucess-color: #00ff00;
   --warning-color: #ff4444;
+  --danger-color: #ff4444;
```
Warning and danger color codes are similar

nyunja committed:
refactor(theme): Use CSS variables for colors in post components

Replaced hardcoded color values throughout the components in the `components/posts` directory with CSS variables defined in `src/app/globals.css`.

This change centralizes the applications color palette, making it easier to maintain and update the theme consistently. It eliminates instances of hardcoded hex values, Tailwind color classes (e.g., `bg-blue-100`, `text-red-400`), and inline styles, ensuring that the components adhere to the established design system.

Affected components:
- PostCreation.jsx
- PostList.jsx
- CommentForm.jsx
- CommentList.jsx
- UserSearch.jsx

nyunja committed:
refactor(theme): Consolidate danger and warning colors

Replaced `--danger-color` with `--warning-color` throughout the application and removed the redundant `--danger-color` from `src/app/globals.css`. This simplifies the color palette and improves maintainability.

Siaka385 reviewed (approved):
looks good
